--- a/src/dyn_crypto.c
+++ b/src/dyn_crypto.c
@@ -126,8 +126,15 @@
 aes_init(void)
 {
     // Initialize contexts
-    aes_encrypt_ctx = (EVP_CIPHER_CTX*) malloc(sizeof(EVP_CIPHER_CTX));
-    aes_decrypt_ctx = (EVP_CIPHER_CTX*) malloc(sizeof(EVP_CIPHER_CTX));
+    aes_encrypt_ctx = EVP_CIPHER_CTX_new();
+    if (aes_encrypt_ctx == NULL) {
+        return DN_ERROR;
+    }
+
+    aes_decrypt_ctx = EVP_CIPHER_CTX_new();
+    if (aes_decrypt_ctx == NULL) {
+        return DN_ERROR;
+    }
 
     EVP_CIPHER_CTX_init(aes_encrypt_ctx);
 
@@ -173,10 +180,8 @@
 rstatus_t
 crypto_deinit(void)
 {
-    EVP_CIPHER_CTX_cleanup(aes_encrypt_ctx);
-    EVP_CIPHER_CTX_cleanup(aes_decrypt_ctx);
-    free(aes_encrypt_ctx);
-    free(aes_decrypt_ctx);
+    EVP_CIPHER_CTX_free(aes_encrypt_ctx);
+    EVP_CIPHER_CTX_free(aes_decrypt_ctx);
     return DN_OK;
 }
 
